### PR TITLE
Automatic Generation of OpenAPI documentation

### PR DESCRIPTION
**Pull request by jtwillis92**
### This pull request changes...

Adds automatic generation of OpenAPI spec documentation for all backend endpoints using the [drf-yasg2](https://pypi.org/project/drf-yasg2/) package.

Additionally, to prevent existing pip packages from being updated to the latest versions when adding this new dependency all of the versions in `Pipfile` were pinned to explicitly comparable versions (~=x.x.x). This circumvents a problem introduced when `PyJWT` is upgraded to v2.0.1 which results in authentication failures.

## FUNCTIONALITY
Addresses https://github.com/raft-tech/TANF-app/issues/564

As is this exposes documentation at a few different URL endpoints, which can be further tuned as desired by the Team. We can either leave all of these in place or only expose one/some of them.

## ReDocs
Available at `/redocs/` - Presents the documentation in a three-panel ReDocs format.

Screenshot from 2021-01-26 17-54-53

## Swagger UI
Available at `/swagger/` - presents the documentation in Swagger UI format

Screenshot from 2021-01-26 17-56-17

## Raw OpenAPI spec
Available at `/swagger.json` or `/swagger.yaml` - presents the OpenAPI specification in either JSON or YAML

## TO TEST

* Navigate to `http://localhost:8080/redocs/` and confirm ReDocs format documentation is available and correctly generated
* Navigate to `http://localhost:8080/swagger/` and confirm Swagger UI style documentation is available and correctly generated
* Navigate to `http://localhost:8080/swagger.[json|yaml]` and confirm OpenAPI spec is generated in the appropriate format.

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  - [x] API documentation must be automatically generated for all exposed endpoints
  - [x] Documentation must meet OpenAPI standards
  - [x] Generated documentation must be convertable to HTML format so that it can be served in a browser
  - [x] Testing Checklist has been run and all tests pass

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [x] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [x] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?
+ [x] If this PR introduces frontend code, is that code documented both inline and overall?
+ [x] If this PR introduces dependencies, are their licenses documented?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

### CODE REVIEW

File: tdrs-backend/tdpservice/urls.py [hunk 1]
```diff
@@ -31,3 +35,24 @@
...
 ] + static(settings.STATIC_URL, document_root=settings.STATIC_ROOT)
+
+
+schema_view = get_schema_view(
+    openapi.Info(
+        title="TDP API",
+        default_version='v1',
+        description="TANF Data Portal API documentation",
+        terms_of_service="TODO",
+        contact=openapi.Contact(email="TODO"),
```
Comment (jtwillis92) [inline tdrs-backend/tdpservice/urls.py:46]: @carltonsmith I wasn't sure what to put here for `terms_of_service` or `contact` but both get exposed in the API documentation UI

**Review by carltonsmith — CHANGES_REQUESTED** [review]
Since we already are using the `/docs` folder, can this be configured so ReDocs uses a different folder? Perhaps `/redocs`?

File: tdrs-backend/tdpservice/urls.py (hunk 1 above)
Comment (carltonsmith) [reply tdrs-backend/tdpservice/urls.py:46]: @jtwillis92 Is it possible to implement if we don't have a terms of service yet? Also, do you know what the email does? Does this email out anything, or just publish the email to the docs?

Comment (jtwillis92) [reply tdrs-backend/tdpservice/urls.py:46]: Yeah those arguments are optional, they just control these links in the UI(s):


Without defining anything they are just not present in the UI / OpenAPI spec.

**Comment by carltonsmith** [comment]
@ADPennington @alexsoble This asks us for links to our email contact and our terms of service. We need to decide who the email contact should be. My assumption would be @ADPennington 

as for the terms of service, we can leave this as `TODO` until we have one, but the link in the docs provided will probably result in a 404. Since this isn't for public consumption I don't think that should hold this up, but I'd like to get your feedback before I put it into QASP Review

**Comment by ADPennington** [comment]
Re: email address, I'd suggest tanfdata@acf.hhs.gov since it's already visible in the build and our data team manages it. Otherwise, if personalized email is necessary, I'm happy to be the POC. 

Also agree re: not holding up things for the terms of service. good to document about broken links until this is ready. @carltonsmith

**Comment by jtwillis92** [comment]
> Also agree re: not holding up things for the terms of service. good to document about broken links until this is ready. @carltonsmith

I can just remove the argument for now and it won't display the ToS link. I'll leave a comment noting that it was removed so we can add it back in the future.

**Comment by carltonsmith** [comment]
> > Also agree re: not holding up things for the terms of service. good to document about broken links until this is ready. @carltonsmith
> 
> I can just remove the argument for now and it won't display the ToS link. I'll leave a comment noting that it was removed so we can add it back in the future.

Ok, that sounds good. You can update the comment I just added accordingly also.

**Review by carltonsmith — APPROVED** [review]
LGTM!

File: tdrs-backend/tdpservice/settings/common.py [hunk 2]
```diff
@@ -80,6 +80,7 @@ class Common(Configuration):
         "django_admin_logs",  # logs for admin site
         "corsheaders",
         "django_extensions",
+        "drf_yasg2",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/settings/common.py:83]: the documentation here (https://github.com/JoelLefkowitz/drf-yasg) says the original package is being maintained again and that it should be used over this one. any reason for utilizing this version? @jtwillis92

Comment (jtwillis92) [reply tdrs-backend/tdpservice/settings/common.py]: Good catch, it does look like they are equivalent at the current time based on the latest release for each but it makes sense to go back to the original package so we can receive future updates.

**Review by ADPennington — APPROVED** [review]
approved! 👍🏾 👍🏾 

@jtwillis92 appreciate the package update (`drf-yasg`) since it's being maintained again. Also great call on pipfile change to help us stay on top of dependency management!

-----
### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
   + [x]  API documentation must be automatically generated for all exposed endpoints
   + [x] Documentation must meet OpenAPI standards
   + [x] Generated documentation must be convertable to HTML format so that it can be served in a browser
  + [x]  Testing Checklist has been run and all tests pass

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?

+ Are code coverage minimums met?
  + [x] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?

### Deliverable 4: Accessible

_n/a_

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

**Pull request merged**